File: lib/services/lead_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/lead_model.dart';
import 'activity_service.dart';

class LeadService {
  final SupabaseClient _supabase = Supabase.instance.client;

  /// Fetch paginated leads, ordered by created_at descending
  Future<List<LeadModel>> getLeads({int page = 0, int pageSize = 20}) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final leadsResponse = await _supabase
        .from('leads')
        .select()
        .order('created_at', ascending: false)
        .range(start, end);

    final List<LeadModel> leads = (leadsResponse as List)
        .map((json) => LeadModel.fromJson(json))
        .toList();

    return leads;
  }

  /// Get a single lead by ID
  Future<LeadModel> getLeadById(String id) async {
    final response = await _supabase
        .from('leads')
        .select()
        .eq('id', id)
        .single();

    return LeadModel.fromJson(response);
  }

  /// Search leads by name, email, or source
  Future<List<LeadModel>> searchLeads(
    String query, {
    int page = 0,
    int pageSize = 20,
  }) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final response = await _supabase
        .from('leads')
        .select()
        .or(
          'first_name.ilike.%$query%,last_name.ilike.%$query%,email.ilike.%$query%,lead_source.ilike.%$query%',
        )
        .order('created_at', ascending: false)
        .range(start, end);

    return (response as List).map((json) => LeadModel.fromJson(json)).toList();
  }

  /// Filter leads by status
  Future<List<LeadModel>> getLeadsByStatus(
    String status, {
    int page = 0,
    int pageSize = 20,
  }) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final response = await _supabase
        .from('leads')
        .select()
        .eq('status', status)
        .order('created_at', ascending: false)
        .range(start, end);

    return (response as List).map((json) => LeadModel.fromJson(json)).toList();
  }

  /// Get leads assigned to a specific user
  Future<List<LeadModel>> getLeadsByAssignee(
    String userId, {
    int page = 0,
    int pageSize = 20,
  }) async {
    final start = page * pageSize;
    final end = start + pageSize - 1;

    final response = await _supabase
        .from('leads')
        .select()
        .eq('assigned_to', userId)
        .order('created_at', ascending: false)
        .range(start, end);

    return (response as List).map((json) => LeadModel.fromJson(json)).toList();
  }

  /// Create a new lead
  Future<LeadModel> addLead(LeadModel lead) async {
    final json = lead.toJson();

    final response = await _supabase
        .from('leads')
        .insert(json)
        .select()
        .single();

    final newLead = LeadModel.fromJson(response);

    // Log Activity: Lead Created
    ActivityService.log(
      relatedType: 'lead',
      relatedId: newLead.id,
      activityType: 'created',
      title: 'Lead Created',
      description: 'New lead added: ${newLead.firstName} ${newLead.lastName}',
      metadata: {'source': newLead.leadSource, 'status': newLead.status},
      organizationId: newLead.organizationId,
    );

    return newLead;
  }

  /// Update an existing lead
  Future<LeadModel> updateLead(LeadModel lead) async {
    final json = lead.toJson();
    json['updated_at'] = DateTime.now().toIso8601String();

    final response = await _supabase
        .from('leads')
        .update(json)
        .eq('id', lead.id)
        .select()
        .single();

    final updatedLead = LeadModel.fromJson(response);

    // Log Activity
    ActivityService.log(
      relatedType: 'lead',
      relatedId: updatedLead.id,
      activityType: 'updated',
      title: 'Lead Updated',
      description: 'Lead info or status changed.',
      metadata: {'status': updatedLead.status},
      organizationId: updatedLead.organizationId,
    );

    return updatedLead;
  }

  /// Delete a lead by ID
  Future<void> deleteLead(String id) async {
    await _supabase.from('leads').delete().eq('id', id);
  }

  /// Converts a Lead to a Contact using the Supabase RPC function, with a fallback if missing.
  Future<String> convertLead(String leadId) async {
    try {
      final response = await _supabase.rpc(
        'convert_lead',
        params: {'lead_uuid': leadId},
      );
      return response as String;
    } catch (e) {
      if (e.toString().contains('Could not find the function') ||
          e.toString().contains('convert_lead')) {
        // Fallback: Perform manual conversion client-side if RPC doesn't exist

        // 1. Fetch lead
        final leadMap = await _supabase
            .from('leads')
            .select()
            .eq('id', leadId)
            .single();
        final String? leadCompanyName = leadMap['company'];
        String? companyId;

        // 2. Intelligent Company Linking/Creation
        if (leadCompanyName != null && leadCompanyName.isNotEmpty) {
          // Check if company already exists
          final existingCompanies = await _supabase
              .from('companies')
              .select('id')
              .ilike('name', leadCompanyName)
              .limit(1);

          if ((existingCompanies as List).isNotEmpty) {
            companyId = existingCompanies[0]['id'];
          } else {
            // Create new company automatically
            final newCompany = await _supabase
                .from('companies')
                .insert({
                  'name': leadCompanyName,
                })
                .select('id')
                .single();
            companyId = newCompany['id'];
          }
        }

        // 3. Insert into contacts
        final contactData = {
          'first_name': leadMap['first_name'],
          'last_name': leadMap['last_name'],
          'email': leadMap['email'],
          'phone': leadMap['phone'],
          'company_name': leadCompanyName ?? '',
          'company_id': companyId,
          'is_customer': false,
          'assigned_to': (leadMap['assigned_to'] != null &&
                  leadMap['assigned_to'].toString().isNotEmpty)
              ? leadMap['assigned_to']
              : null,
          'notes': leadMap['notes'],
          'created_from_lead': true,
          'source_lead_id': leadId,
        };

        final contactRes = await _supabase
            .from('contacts')
            .insert(contactData)
            .select('id')
            .single();
        final contactId = contactRes['id'] as String;

        // 4. Update lead
        await _supabase
            .from('leads')
            .update({
              'status': 'converted',
              'converted_contact_id': contactId,
              'converted_at': DateTime.now().toIso8601String(),
              'updated_at': DateTime.now().toIso8601String(),
            })
            .eq('id', leadId);

        // Log Activity: Lead Converted
        ActivityService.log(
          relatedType: 'lead',
          relatedId: leadId,
          activityType: 'converted',
          title: 'Lead Converted',
          description: 'Lead converted to contact.',
          metadata: {'contact_id': contactId, 'company_id': companyId},
          organizationId: leadMap['organization_id'],
        );

        // Also log for the new contact
        ActivityService.log(
          relatedType: 'contact',
          relatedId: contactId,
          activityType: 'created',
          title: 'Contact Created from Lead',
          description: 'Contact established via lead conversion.',
          metadata: {'source_lead_id': leadId},
          organizationId: leadMap['organization_id'],
        );

        return contactId;
      }
      rethrow;
    }
  }

  /// Get lead pipeline statistics
  Future<Map<String, dynamic>> getStats() async {
    final response = await _supabase
        .from('leads')
        .select('id, status, estimated_value, created_at');
    final List<dynamic> data = response as List<dynamic>;

    final now = DateTime.now();
    final thirtyDaysAgo = now.subtract(const Duration(days: 30));

    int total = data.length;
    int newLeads = data.where((l) => l['status'] == 'new_lead').length;
    int contacted = data.where((l) => l['status'] == 'contacted').length;
    int interested = data.where((l) => l['status'] == 'interested').length;
    int qualified = data.where((l) => l['status'] == 'qualified').length;
    int lost = data.where((l) => l['status'] == 'lost').length;
    int converted = data.where((l) => l['status'] == 'converted').length;
    int newThisMonth = data.where((l) {
      final createdAt = DateTime.tryParse(l['created_at'] ?? '');
      return createdAt != null && createdAt.isAfter(thirtyDaysAgo);
    }).length;

    double pipelineValue = data
        .where((l) => l['status'] != 'lost' && l['status'] != 'converted')
        .fold(
          0.0,
          (sum, l) => sum + ((l['estimated_value'] as num?)?.toDouble() ?? 0.0),
        );

    return {
      'total': total,
      'new_leads': newLeads,
      'contacted': contacted,
      'interested': interested,
      'qualified': qualified,
      'lost': lost,
      'converted': converted,
      'pipeline_value': pipelineValue,
      'new_this_month': newThisMonth,
    };
  }
}

File: backend/src/services/LeadService.js
const db = require('../config/db');

/**
 * LeadService - Direct PostgreSQL CRUD operations for the leads table.
 * Mirrors the Supabase schema_v2.sql `leads` table structure.
 */
class LeadService {

    /**
     * Fetch paginated leads ordered by created_at DESC
     */
    static async getLeads({ page = 0, pageSize = 20, status, source } = {}) {
        const offset = page * pageSize;
        let query = 'SELECT * FROM leads';
        const params = [];
        const conditions = [];

        if (status) {
            params.push(status);
            conditions.push(`status = $${params.length}`);
        }
        if (source) {
            params.push(source);
            conditions.push(`lead_source = $${params.length}`);
        }

        if (conditions.length > 0) {
            query += ' WHERE ' + conditions.join(' AND ');
        }

        query += ' ORDER BY created_at DESC';
        params.push(pageSize);
        query += ` LIMIT $${params.length}`;
        params.push(offset);
        query += ` OFFSET $${params.length}`;

        const result = await db.query(query, params);
        return result.rows;
    }

    /**
     * Get single lead by ID
     */
    static async getLeadById(id) {
        const result = await db.query(
            'SELECT * FROM leads WHERE id = $1',
            [id]
        );
        if (result.rows.length === 0) {
            throw new Error('Lead not found');
        }
        return result.rows[0];
    }

    /**
     * Search leads by name, email, or source
     */
    static async searchLeads(query, { page = 0, pageSize = 20 } = {}) {
        const offset = page * pageSize;
        const searchTerm = `%${query}%`;
        const result = await db.query(
            `SELECT * FROM leads
             WHERE first_name ILIKE $1
                OR last_name ILIKE $1
                OR email ILIKE $1
                OR lead_source ILIKE $1
             ORDER BY created_at DESC
             LIMIT $2 OFFSET $3`,
            [searchTerm, pageSize, offset]
        );
        return result.rows;
    }

    /**
     * Create a new lead
     */
    static async createLead(data) {
        const {
            first_name,
            last_name,
            email,
            phone,
            lead_source,
            status = 'new_lead',
            assigned_to,
            notes,
            estimated_value,
        } = data;

        const result = await db.query(
            `INSERT INTO leads
                (first_name, last_name, email, phone, lead_source, status, assigned_to, notes, estimated_value)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
             RETURNING *`,
            [first_name, last_name || '', email, phone, lead_source, status, assigned_to || null, notes, estimated_value]
        );
        return result.rows[0];
    }

    /**
     * Update an existing lead
     */
    static async updateLead(id, data) {
        const {
            first_name,
            last_name,
            email,
            phone,
            lead_source,
            status,
            assigned_to,
            notes,
            estimated_value,
        } = data;

        const result = await db.query(
            `UPDATE leads SET
                first_name = COALESCE($1, first_name),
                last_name = COALESCE($2, last_name),
                email = COALESCE($3, email),
                phone = COALESCE($4, phone),
                lead_source = COALESCE($5, lead_source),
                status = COALESCE($6, status),
                assigned_to = COALESCE($7, assigned_to),
                notes = COALESCE($8, notes),
                estimated_value = COALESCE($9, estimated_value),
                updated_at = NOW()
             WHERE id = $10
             RETURNING *`,
            [first_name, last_name, email, phone, lead_source, status, assigned_to || null, notes, estimated_value, id]
        );

        if (result.rows.length === 0) {
            throw new Error('Lead not found');
        }
        return result.rows[0];
    }

    /**
     * Delete a lead by ID
     */
    static async deleteLead(id) {
        const result = await db.query(
            'DELETE FROM leads WHERE id = $1 RETURNING id',
            [id]
        );
        if (result.rows.length === 0) {
            throw new Error('Lead not found');
        }
        return { deleted: true, id: result.rows[0].id };
    }

    /**
     * Convert a lead to a contact (atomic transaction)
     * Replicates the convert_lead RPC function from schema_v2.sql
     */
    static async convertLead(leadId) {
        const client = await require('../config/db');

        // We need a raw pg client for transactions, so we use the pool directly
        const { Pool } = require('pg');
        require('dotenv').config();
        const pool = new Pool({
            connectionString: process.env.DATABASE_URL,
            ssl: { rejectUnauthorized: false }
        });
        const txClient = await pool.connect();

        try {
            await txClient.query('BEGIN');

            // 1. Lock and fetch the lead
            const leadResult = await txClient.query(
                'SELECT * FROM leads WHERE id = $1 FOR UPDATE',
                [leadId]
            );

            if (leadResult.rows.length === 0) {
                throw new Error('Lead not found');
            }

            const lead = leadResult.rows[0];

            if (lead.status === 'converted') {
                throw new Error('Lead is already converted');
            }

            // 2. Insert into contacts
            const contactResult = await txClient.query(
                `INSERT INTO contacts
                    (first_name, last_name, email, phone, assigned_to, notes, created_from_lead, source_lead_id, is_customer)
                 VALUES ($1, $2, $3, $4, $5, $6, TRUE, $7, FALSE)
                 RETURNING id`,
                [lead.first_name, lead.last_name, lead.email, lead.phone, lead.assigned_to, lead.notes, lead.id]
            );

            const contactId = contactResult.rows[0].id;

            // 3. Update lead as converted
            await txClient.query(
                `UPDATE leads
                 SET status = 'converted',
                     converted_at = NOW(),
                     converted_contact_id = $1,
                     updated_at = NOW()
                 WHERE id = $2`,
                [contactId, leadId]
            );

            await txClient.query('COMMIT');

            return { contactId, leadId };
        } catch (error) {
            await txClient.query('ROLLBACK');
            throw error;
        } finally {
            txClient.release();
            await pool.end();
        }
    }

    /**
     * Get lead pipeline statistics (count per status)
     */
    static async getStats() {
        const result = await db.query(`
            SELECT
                COUNT(*) AS total,
                COUNT(*) FILTER (WHERE status = 'new_lead') AS new_leads,
                COUNT(*) FILTER (WHERE status = 'contacted') AS contacted,
                COUNT(*) FILTER (WHERE status = 'interested') AS interested,
                COUNT(*) FILTER (WHERE status = 'qualified') AS qualified,
                COUNT(*) FILTER (WHERE status = 'lost') AS lost,
                COUNT(*) FILTER (WHERE status = 'converted') AS converted,
                COALESCE(SUM(estimated_value) FILTER (WHERE status != 'lost' AND status != 'converted'), 0) AS pipeline_value,
                COUNT(*) FILTER (WHERE created_at >= NOW() - INTERVAL '30 days') AS new_this_month
            FROM leads
        `);
        return result.rows[0];
    }

    /**
     * Get leads assigned to a specific user
     */
    static async getLeadsByAssignee(userId, { page = 0, pageSize = 20 } = {}) {
        const offset = page * pageSize;
        const result = await db.query(
            `SELECT * FROM leads
             WHERE assigned_to = $1
             ORDER BY created_at DESC
             LIMIT $2 OFFSET $3`,
            [userId, pageSize, offset]
        );
        return result.rows;
    }
}

module.exports = LeadService;
